packages/server/src/cache/leaderboard.ts

```diff
@@ -0,0 +1,225 @@
+    leaderboard: result
+  }
+}
+
+export const getScore = async ({
+  id
+}: Pick<User, 'id'>): Promise<null | { score: UserInfo['score'], globalPlace: number, divisionPlace: number }> => {
+  const redisResult = await client.hget('score-positions', id)
+  if (redisResult === null) {
+    return null
+  }
+  const split = redisResult.split(',')
+  return {
+    score: parseInt(split[0]),
+    globalPlace: parseInt(split[1]),
+    divisionPlace: parseInt(split[2])
+  }
+}
+
+export const getChallengeInfo = async ({ ids }: { ids: Challenge['id'][] }): Promise<ExternalChallengeInfo[]> => {
+  if (ids.length === 0) {
+    return []
+  }
+  const redisResult = await client.hmget('challenge-info', ...ids)
+  return redisResult.map((info) => {
+    if (info === null) {
+      return {
+        score: null,
+        solves: null
+      }
+    }
+    const split = info.split(',')
+    return {
+      score: parseInt(split[0]),
+      solves: parseInt(split[1])
+    }
+  })
+}
+
+client.defineCommand('rctfSetGraph', {
+  lua: loadScript('set-graph')
+})
+export type scriptSetGraph = (keys: number, args: string[]) => Promise<void>
+
+export const setGraph = async ({ graphLeaderboards }: Pick<WorkerResponse, 'graphLeaderboards'>): Promise<void> => {
+  let lastSample = 0
+  const userPoints = new Map<string, number[]>()
+  for (let entryIdx = 0; entryIdx < graphLeaderboards.length; entryIdx++) {
+    const graphEntry = graphLeaderboards[entryIdx]
+    if (graphEntry.sample > lastSample) {
+      lastSample = graphEntry.sample
+    }
+    for (let userIdx = 0; userIdx < graphEntry.userInfos.length; userIdx++) {
+      const userInfo = graphEntry.userInfos[userIdx]
+      const key = `graph:${userInfo.id}`
+      const userPoint = userPoints.get(key)
+      if (userPoint === undefined) {
+        userPoints.set(key, [graphEntry.sample, userInfo.score])
+      } else {
+        userPoint.push(graphEntry.sample, userInfo.score)
+      }
+    }
+  }
+  if (userPoints.size === 0) {
+    return
+  }
+  const keys = Array.from(userPoints.keys())
+  const values = Array.from(userPoints.values())
+  await client.rctfSetGraph(1 + keys.length, [
+    'graph-update',
+    ...keys,
+    lastSample.toString(),
+    JSON.stringify(values)
+  ])
+}
+
+client.defineCommand('rctfGetGraph', {
+  numberOfKeys: 2,
+  lua: loadScript('get-graph')
+})
+export type scriptGetGraph = (leaderboardKey: string, leaderboardUpdateKey: string, maxTeams: string) => Promise<string | null>
+
+export const getGraph = async ({
+  division,
+  maxTeams
+}: { division: string, maxTeams: number }): Promise<ExternalGraphEntry[]> => {
+  const redisResult = await client.rctfGetGraph(
+    getLeaderboardKey(division),
+    'leaderboard-update',
+    maxTeams.toString()
+  )
+  if (redisResult === null) {
+    return []
+  }
+  const parsed = JSON.parse(redisResult) as string[]
+  const lastUpdate = parseInt(parsed[0])
+  const latest = parsed[1]
+  const graphData = parsed[2]
+  const result = []
```

Comment:
This should be typed.
```suggestion
  const result: ExternalGraphEntry[] = []
```